pkg/apis/ceph.rook.io/v1/scc.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"bytes"
+	_ "embed"
+
+	secv1 "github.com/openshift/api/security/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+const (
+	// Determines how far into the stream the decoder will look to figure out whether
+	// this is a JSON stream or YAML
+	walkThroughBufferSize = 10
+)
+
+var (
+	//go:embed template/scc.yaml
+	securityContextConstraintsFilePath []byte
+	//go:embed template/scc-csi.yaml
+	securityContextConstraintsCSIFilePath []byte
+)
+
+// NewSecurityContextConstraints returns a new SecurityContextConstraints for Rook-Ceph to run on
+// OpenShift.
+func NewSecurityContextConstraints() (*secv1.SecurityContextConstraints, error) {
```

Comment:
Creating the SCCs in code is a security concern because that means the caller has enough privileges to create SCCs, which really is full admin access because you can grant permissions for anything. I don't think it's Rook's job to support this scenario. If some other component such as downstream needs to do it, then it's downstream's problem.

Reply:
I agree with that point as well.

Reply:
As an idea, we could have the `kubectl-ceph` plugin create SCC in a command if we so desire.

Reply:
I don't see how it's a concern honestly, IMO there is nothing wrong with providing the SCC as code as well YAML. If the caller has the right privileges then good it can use it. Provided as code is just another delivery method, people can choose to use Helm, the YAML, or the code if they want to.

Reply:
If someone gets access to the Rook pod, they could more easily gain access to other openshift components by being able to create/modify SCCs that grant more permissions.

Reply:
Rook is not creating the permissions. It's just exposing via its API what should be the right SCC to inject.

Reply:
But why add the complexity of doing it from rook code? The OCS code is already simple, why not keep it there?

Reply:
The main reason is that removes some burden for everyone. OCS can just pull it without having to know anything about it and we don't need to push changes in OCS.

Reply:
It's also increasing burden of upstream rook to maintain this code and also for users to have a separate scc.yaml instead of including in operator-openshift.yaml. Was there a requirement for a change? Maybe that's what I'm still missing. The approach without this PR just seems simplest.

Reply:
Alternatively, to keep it simple, the code version of creating the SCCs could be moved from OCS to Rook. This would allow more flexibility for dynamic SCC settings for anyone calling Rook as an API, and we could keep the upstream SCCs as yamls.

Reply:
> Alternatively, to keep it simple, the code version of creating the SCCs could be moved from OCS to Rook. This would allow more flexibility for dynamic SCC settings for anyone calling Rook as an API, and we could keep the upstream SCCs as yamls.

I was trying to avoid that since it would really duplicate the code and the yamls. This implementation is cleaner IMO since the code is based on the yamls content. The code just wraps the yaml and exports it in our API.